cookbooks/bcpc-hadoop/recipes/oozie.rb

```diff
@@ -65,61 +56,73 @@
+    not File.exists?("#{OOZIE_SERVER_PATH}/webapps/oozie.war") or
+    File.mtime("#{OOZIE_CLIENT_PATH}/libext/") >
+      File.mtime("#{OOZIE_SERVER_PATH}/webapps/oozie.war")
+  }
 end
 
-bash "make_shared_libs_dir" do
-  code <<EOH
-  hdfs dfs -mkdir -p #{HDFS_URL}/user/oozie/share/ && \
-  hdfs dfs -chown -R oozie #{HDFS_URL}/user/oozie/
-EOH
+bash "make_oozie_user_dir" do
+  code <<-EOH
+    hdfs dfs -mkdir -p #{HDFS_URL}/user/oozie && \
+    hdfs dfs -chown -R oozie #{HDFS_URL}/user/oozie
+  EOH
   user "hdfs"
-  not_if "hdfs dfs -test #{HDFS_URL}/user/oozie/share/", :user => "hdfs"
-end
-
-#bash "untar_sharelib_files" do
-#  code <<EOH
-#  cd #{OOZIE_CLIENT_PATH}
-#  tar -xzf oozie-sharelib.tar.gz
-#EOH
-#  not_if { ::File.directory?("#{OOZIE_CLIENT_PATH}/share") }
-#end
-#
-#bash "oozie_create_shared_libs" do
-#  share_dir_url="#{HDFS_URL}/user/oozie/share/"
-#  code "#{OOZIE_CLIENT_PATH}/bin/oozie-setup.sh sharelib create -fs #{HDFS_URL} -locallib /usr/hdp/current/oozie-client/share"
-#  user "oozie"
-#  not_if "hdfs dfs -test -d #{HDFS_URL}/user/oozie/share/lib", :user => "hdfs"
-#end
-
-bash "oozie_update_shared_libs" do
-  share_dir_url="#{HDFS_URL}/user/oozie/share/"
-  #code "#{OOZIE_LIB_PATH}/bin/oozie-setup.sh sharelib update -fs #{HDFS_URL}"
-  code "#{OOZIE_CLIENT_PATH}/bin/oozie-setup.sh sharelib upgrade -fs #{HDFS_URL}"
+  not_if "hdfs dfs -test -d #{HDFS_URL}/user/oozie", :user => "hdfs"
+end
+
+bash "oozie_create_shared_libs" do
+  code "#{OOZIE_CLIENT_PATH}/bin/oozie-setup.sh sharelib create -fs #{HDFS_URL} -locallib #{OOZIE_SHARELIB_TARBALL_PATH}"
   user "oozie"
-  not_if "sudo -u hdfs hdfs dfs -test -d #{HDFS_URL}/user/oozie/share/lib", :user => "hdfs"
-  only_if "echo 'test'|sudo -u hdfs hdfs dfs -copyFromLocal - /tmp/oozie-test"
-  notifies :run,"bash[delete-oozie-temp-file]",:immediately
-  #not_if { require 'time'
-  # hdfs_mtime=`hdfs dfs -stat #{share_dir_url}`.strip
-  # Time.parse("#{hdfs_mtime} UTC") >
-  # File.mtime("#{OOZIE_CLIENT_PATH}/oozie-sharelib.tar.gz") }
+  not_if {
+    require 'digest'
+    Digest::MD5.hexdigest(File.read(OOZIE_SHARELIB_TARBALL_PATH)) == node[:bcpc][:hadoop][:oozie][:sharelib_checksum]
+  } 
+  only_if "echo 'test'| hdfs dfs -copyFromLocal - /tmp/oozie-test && hdfs dfs -rm -skipTrash /tmp/oozie-test", :user => "hdfs"
+  notifies :run, "ruby_block[update_sharelib_checksum]", :immediately
 end
 
-bash "delete-oozie-temp-file" do
-  code <<-EOH
-    hdfs dfs -rm /tmp/oozie-test
-  EOH
-  user "hdfs"
+ruby_block "update_sharelib_checksum" do
+  block do
+    require 'digest'
+    node.set[:bcpc][:hadoop][:oozie][:sharelib_checksum] = 
+      Digest::MD5.hexdigest(File.read(OOZIE_SHARELIB_TARBALL_PATH))
+  end
+  action :nothing
+  notifies :run, "ruby_block[notify_sharelib_update]", :immediately
+end
+
+update_oozie_sharelib = lambda do |host|
+  begin
+    oozie_status=`sudo -u oozie oozie admin -oozie http://#{host}:11000/oozie -status 2>&1`
+    if /NORMAL/ =~ oozie_status && $?.to_i
```

Comment:
$? has the methods "success?" and "exitstatus" for this purpose

(as with perl, taking $? directly as an integer can have surprising results, because the exit code is stored in the high bits of a 16 bit integer)